runtime/near-vm-logic/src/logic.rs

```diff
@@ -812,6 +812,68 @@ impl<'a> VMLogic<'a> {
         self.internal_write_register(register_id, value_hash)
     }
 
+    /// Compute multiexponentiation for (sign, exp, base), serialized at bytes in LE notation.
+    ///
+    /// # Errors
+    ///
+    /// If `value_len + value_ptr` points outside the memory or the registers use more memory than
+    /// the limit with `MemoryAccessViolation`.
+    ///
+    /// If `value_len-4 % 97 != 0`
+    /// # Cost
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + alt_bn128_pairing_base + alt_bn128_pairing_byte * num_bytes`
+    pub fn alt_bn128_g1_multiexp(
+        &mut self,
+        value_len: u64,
+        value_ptr: u64,
+        register_id: u64,
+    ) -> Result<()> {
+        fn weight(data: &[u8]) -> usize {
+            let mut log2 = 0;
+            let mut hamming = 0;
+            for i in 0..data.len() {
+                for j in 0..8 {
+                    if data[i] >> j == 1 {
+                        log2 = i * 8 + j;
+                        hamming += 1;
+                    }
+                }
+            }
+            log2 + hamming + 1
+        }
+
+        self.gas_counter.pay_base(alt_bn128_g1_multiexp_base)?;
+        let value = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
+
+        let mut w_acc = 0;
```

Comment:
If we decide to keep implementation of atl_bn128 math in the implementor of `External` trait then we should not pass data in the serialized format to `External`. Instead we should do deserialization here, and then have a strongly typed interface of `External`. Also, it seems like we are doing manual deserialization here, and I suspect manual serialization in `near-sdk-rs`. Instead we should be using borsh.io

Reply:
Borsh is not implemented for parity-bn types. We can fork parity-bn or use wrapped types. What way is the best?

Reply:
Wrap them in types would work. But we should also consider other Rust cryptographic libraries, we also need to make sure parity-bn compiles to Wasm. @abacabadabacaba WDYT? Is parity-bn a suitable library, are there alternatives?

Reply:
parity-bn is forked now, [zeropool-bn](https://github.com/zeropoolnetwork/bn) supports borsh.